python/ray/experimental/data/datasource/binary_datasource.py

```diff
@@ -0,0 +1,20 @@
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    import pyarrow
+
+from ray.experimental.data.datasource.file_based_datasource import (
+    FileBasedDatasource)
+
+
+class BinaryDatasource(FileBasedDatasource):
+    """Binary datasource, for reading and writing binary files.
+
+    Examples:
+        >>> source = BinaryDatasource()
+        >>> ray.data.read_datasource(source, paths="/path/to/dir").take()
+        ... {"a": 1, "b": "foo"}
+    """
+
+    def _read_file(self, f: "pyarrow.NativeFile", **arrow_reader_args):
+        return f.readall()
```

Comment:
Add read option to include the path here?

Reply:
Yep that's what I'm doing, I was hoping to find a way to get the path off of the `pyarrow.NativeFile` but it doesn't look like that's exposed.